osu.Game/IO/HardLinkHelper.cs

```diff
@@ -111,22 +125,50 @@ public static int GetFileLinkCount(string filePath)
         [DllImport("Kernel32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
         public static extern bool CreateHardLink(string lpFileName, string lpExistingFileName, IntPtr lpSecurityAttributes);
 
-        [DllImport("kernel32.dll", SetLastError = true, CharSet = CharSet.Auto)]
+        // https://learn.microsoft.com/windows/win32/fileio/obtaining-a-handle-to-a-directory
+        // The flag is required when opening directory as a handle. It's not accepted by File.OpenHandle.
+        public const uint FILE_FLAG_BACKUP_SEMANTICS = 0x02000000;
+
+        [DllImport("kernel32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
         private static extern SafeFileHandle CreateFile(
             string lpFileName,
-            [MarshalAs(UnmanagedType.U4)] FileAccess dwDesiredAccess,
-            [MarshalAs(UnmanagedType.U4)] FileShare dwShareMode,
+            FileAccess dwDesiredAccess,
+            FileShare dwShareMode,
             IntPtr lpSecurityAttributes,
-            [MarshalAs(UnmanagedType.U4)] FileMode dwCreationDisposition,
-            [MarshalAs(UnmanagedType.U4)] FileAttributes dwFlagsAndAttributes,
+            FileMode dwCreationDisposition,
+            FileAttributes dwFlagsAndAttributes,
             IntPtr hTemplateFile);
 
-        [DllImport("kernel32.dll", SetLastError = true)]
-        private static extern bool GetFileInformationByHandle(SafeFileHandle handle, out ByHandleFileInformation lpFileInformation);
+        public const uint FILE_SUPPORTS_HARD_LINKS = 0x00400000;
+        public const uint FILE_SUPPORTS_BLOCK_REFCOUNTING = 0x08000000;
+
+        [DllImport("Kernel32.dll", CharSet = CharSet.Unicode)]
+        private static extern bool GetVolumeInformationByHandle(
+            SafeFileHandle hFile,
+            IntPtr lpVolumeNameBuffer,
+            uint nVolumeNameSize,
+            out uint lpVolumeSerialNumber,
+            out uint lpMaximumComponentLength,
+            out uint lpFileSystemFlags,
+            IntPtr lpFileSystemNameBuffer,
+            uint nFileSystemNameSize);
```

Comment:
> * `lp{VolumeSerialNumber,MaximumComponentLength,FileSystemFlags}` should be _pointers_?

`out X` or `ref X` are equivalent to `X*`.

> * `lpVolumeNameBuffer` should be an `out` param?

It's an (optional) pointer to start of a buffer, not a single variable. Using `out` is not complete here. The correct approach is to allocate a `Span<char>` and take a `ref` to its start element.

Since it's optional and not used by the caller, it's simpler to just pass `(char*)null`, a.k.a. `IntPtr.Zero` under safe syntax. This is the common pattern used at https://github.com/dotnet/runtime/blob/1e48f9ebbfd11ac31e07d03761db991d564b2505/src/libraries/Common/src/Interop/Windows/Kernel32/Interop.ReadFile_IntPtr.cs#L17 .